sound/soc/sof/intel/hda-pcm.c

```diff
@@ -95,6 +95,53 @@ static inline u32 get_bits(struct snd_sof_dev *sdev, int sample_bits)
 	}
 };
 
+/* There are two dma modes in dsp:
+ * (1) host dma and link dma work in decouple mode for SOF + HDA
+ * (2) host dma and GP dma work in couple mode for SOF + I2S
+ * The difference is that in decopule mode, the host dma is set by dai
+ * frontend and link dma is set by dai backend and the setting maybe
+ * different while in couple mode, host dma and GP dma share the same
+ * dma setting.
+ */
+static inline int get_be_dma_channel(struct snd_sof_dev *sdev,
+				     struct snd_pcm_substream *substream,
+				     struct sof_ipc_stream_params *ipc_params)
+{
+	struct snd_soc_pcm_runtime *fe = substream->private_data;
+	struct hdac_ext_stream *be_stream = NULL;
+	struct snd_soc_dpcm *dpcm;
+	int direction = substream->stream;
+
+	/* travel BE clients to get BE stream with link dma id */
+	list_for_each_entry(dpcm, &fe->dpcm[direction].be_clients, list_be) {
+		struct snd_soc_pcm_runtime *be = dpcm->be;
+
+		be_stream = snd_soc_dai_get_dma_data(be->cpu_dai, substream);
+		if (be_stream)
+			break;
+	}
+
+	if (be_stream) {
+		/* link dma id is derived from stream_tag in decouple mode*/
+		ipc_params->be_dma_ch = be_stream->hstream.stream_tag - 1;
+	} else {
+		/* the setting of GP dma is same as host dma in default */
+		ipc_params->be_dma_ch = ipc_params->host_dma_ch;
```

Comment:
this makes no sense, sorry. the GP dma channels are not selected by the driver but chosen inside of the firmware.

Reply:
yes, but now this is the algorithm in FW for HDA. For GP-DMA, this field is unused, so no matter what it is.

Reply:
I could refine the comment, how about it?